grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java

```diff
@@ -472,6 +495,22 @@ private void prepareHeaders(Compressor compressor, Metadata metadata, long remai
         ctx.updateRequest(newReq);
     }
 
+    private void closeWhenListenerThrows(Throwable t) {
+        if (needsDirectInvocation()) {
+            close(GrpcStatus.fromThrowable(t), new Metadata());
+        } else {
+            execute(() -> close(GrpcStatus.fromThrowable(t), new Metadata()));
+        }
+    }
+
+    private void closeWhenEos(Status status, Metadata metadata) {
+        if (needsDirectInvocation()) {
+            close(status, metadata);
+        } else {
+            execute(() -> close(status, metadata));
+        }
+    }
+
     private void close(Status status, Metadata metadata) {
```

Comment:
This is not your fault but I found out while reviewing this that we were calling this method not in the eventLoop.
https://github.com/line/armeria/pull/4851/files#diff-39112577712e2be6800f84834abc33a690eed333cd96581b777e1e1cd1ccdadaR234
Could you also fix that, please?

Reply:
Hmm. I looked around more and I think it's fine if we keep running `start()` on the caller's thread for the time being. I added some comments about visibility.

Reply:
Yeah, I also think it's fine. 👍 
Thanks for the clarification and the change. 😄

Reply:
Oops, I realized that we cannot use the `ctx.eventLoop()` because it's not initialized when we call `close` in the `start` method.
I've also checked that it's okay to call close in the caller thread because the `closed` field is visible to other threads because `execute()` method is called by the caller thread.
So let's revert it and add a small comment about:
```
// ctx.eventLoop() might not be initialized so we call close directly from the caller thread.
// It's okay because the visibility of the closed field to other threads is guaranteed.
```

Reply:
Yea I realized that - I'm undoing the change.

One piece that I think we should probably change is that we should fail fast when deadline is expired instead of going through the trouble of trying to send a request (e.g., [this line](https://github.com/line/armeria/pull/4851/files#diff-39112577712e2be6800f84834abc33a690eed333cd96581b777e1e1cd1ccdadaR239) should have return following it) There is also a problem with leaking resources (i.e., req stream) when [compressor is not found](https://github.com/line/armeria/pull/4851/files#diff-39112577712e2be6800f84834abc33a690eed333cd96581b777e1e1cd1ccdadaL205).

I think we should unify these two early returns (when compressor is not found; when deadline is expired) and do the following:
 - fail fast w/o calling into `initContextAndExecuteWithFallback` and initiating a request
 - properly abort `req` and cancel `upstream` subscription (i.e., just call close in both cases)
 
 @minwoox What do you think about it? I tried it locally and it almost works except for I need to also add `if (closed)` checks within `cancel` and `sendMessage` to account for vanilla grpc doing starc/sendMessage/cancel pattern unconditionally.

Reply:
Thanks a lot for the analysis and I don't know why we don't do the early return there. 😅 
@line/dx Do you have any idea about this?
It's added by this PR: https://github.com/line/armeria/pull/2284

Reply:
I added the logic to do early return on deadline expired. I also promoted `closed` to be volatile - I think it's the most safe thing to do. For now we simply ignore the user action on `ClientCall` after it's closed, which is the behavior that ClientCalls.java utilities seem to expect. Let me know what do you think about this structure.

Reply:
Yea I think you're right that executors (user-configured and event-loop) should probably see updates from one another. It's the callers thread driving ClientCall (calling in `sendMessage`, `halfClose`, etc) that I think is out of luck. 

Even though `ClientCall` warrants some form of external synchronization >

(from their javadoc)

> Methods are guaranteed to be non-blocking. Not thread-safe except for request, which may be called from any thread.

There is no synchronization between event-loop (or CallOptions.executor for that matter) and the caller thread. To be fair, there was never one in place (see cancel calls that checks for `closed`) so maybe it's ok that we don't make it immediately better - but I want to make sure we don't make it immediately worse.

Reply:
Okay - I reworked it:

 - added a commentary to explain thread-safety of `close`
 - added early return for when deadline is expired
 - we now properly recycle resources when compressor is not found

@minwoox PTAL